src/fides/api/ctl/sql_models.py

```diff
@@ -313,3 +315,64 @@ class SystemScans(Base):
     for _, sql_model in sql_model_map.items()
     if hasattr(sql_model, "is_default")
 ]
+
+
+class AllowedTypes(str, EnumType):
+    """Allowed types for custom field."""
+
+    string = "string"
+    string_list = "string[]"
+
+
+class ResourceTypes(str, EnumType):
+    """Resource types that can use custom fields."""
+
+    system = "system"
+    data_use = "data use"
+    data_category = "data category"
+    data_subject = "data subject"
+
+
+class CustomFieldDefinition(Base):
+    """Defines custom metadata for resources."""
+
+    __tablename__ = "plus_custom_field_definition"
+
+    fides_key = Column(String, index=True, unique=True)
+    name = Column(String, index=True, nullable=False)
+    description = Column(String)
+    field_type = Column(
+        EnumColumn(AllowedTypes),
+        nullable=False,
+    )
+    allow_list = Column(
+        ARRAY(String)
+    )  # Ulitmately this will come the the CustomList table
+    resource_type = Column(EnumColumn(ResourceTypes), nullable=False)
+    field_definition = Column(String, index=True)
+    custom_field = relationship(
+        "CustomField",
+        back_populates="plus_custom_field_definition",
+        cascade="delete, delete-orphan",
+    )
+    UniqueConstraint("name", "resource_type")
+
+
+class CustomField(Base):
```

Comment:
my inclination tells me to stick with the "soft" foreign key reference and enforce integrity in the python code, in order to (as you mention) keep things flexible going forward. it would allow us to apply custom fields to more resource types going forward, without needing to create new tables and deal with database migrations.

that being said, there's a risk i'm trying to over-genericize the use case here - maybe we can get some input from product (cc @mfbrown) on how likely it'd be that we'd want to eventually apply custom fields to other resource types beyond what we've scoped out here initially? to me it feels like a good future-proofing, but i'm not sure if i'm just going overboard with this.

Reply:
I agree that there might be cases for custom fields on other types and if we can plan for that, we should. Right off the bat, I can plausibly see a world where someone needs a custom field on organizations, for example, which we keep pretty light today, but as we expand to support multiple orgs, could be useful.

Just one example, but I do see the application here growing over time.